nifi-nar-bundles/nifi-jms-bundle/nifi-jms-processors/src/main/java/org/apache/nifi/jms/processors/ConsumeJMS.java

```diff
@@ -78,22 +77,24 @@
      */
     @Override
     protected void rendezvousWithJms(ProcessContext context, ProcessSession processSession) throws ProcessException {
-        final JMSResponse response = this.targetResource.consume();
-        if (response != null){
-            FlowFile flowFile = processSession.create();
-            flowFile = processSession.write(flowFile, new OutputStreamCallback() {
-                @Override
-                public void process(final OutputStream out) throws IOException {
-                    out.write(response.getMessageBody());
-                }
-            });
-            Map<String, Object> jmsHeaders = response.getMessageHeaders();
-            flowFile = this.updateFlowFileAttributesWithJmsHeaders(jmsHeaders, flowFile, processSession);
-            processSession.getProvenanceReporter().receive(flowFile, context.getProperty(DESTINATION).evaluateAttributeExpressions().getValue());
-            processSession.transfer(flowFile, REL_SUCCESS);
-        } else {
-            context.yield();
-        }
+        this.targetResource.consume(response -> {
+            if (response != null) {
+                FlowFile flowFile = processSession.create();
+                flowFile = processSession.write(flowFile, new OutputStreamCallback() {
+                    @Override
+                    public void process(final OutputStream out) throws IOException {
+                        out.write(response.getMessageBody());
+                    }
+                });
+                Map<String, Object> jmsHeaders = response.getMessageHeaders();
+                flowFile = this.updateFlowFileAttributesWithJmsHeaders(jmsHeaders, flowFile, processSession);
+                processSession.getProvenanceReporter().receive(flowFile,
+                        context.getProperty(DESTINATION).evaluateAttributeExpressions().getValue());
+                processSession.transfer(flowFile, REL_SUCCESS);
```

Comment:
Question: doesn't the session need to be explicitly committed here, before the message gets acknowledged, in order to complete the chain of custody?  Otherwise I believe there is a small window where if NiFi crashed before session is committed by scheduler, the acknowledged message can be lost.